Add spider thermostat

peternijssen opened the pull request:
## Description:

This adds the [Spider](http://www.ithodaalderop.nl/spider-thermostaat) thermostat from Itho Daalderop. For now you can see the actual temperature and set temperature. It also allows you to change the temperature.

Any feedback is welcome to improve anything. (This is my first Python code ever, based on many other parts of code of HA).

**Pull request in [home-assistant.github.io](https://github.com/home-assistant/home-assistant.github.io) with documentation (if applicable):** home-assistant/home-assistant.github.io#5805

## Example entry for `configuration.yaml` (if applicable):
```yaml
spider:
  username: username
  password: password
```

## Checklist:
  - [x] The code change is tested and works locally.
  - [x] Local tests pass with `tox`. **Your PR cannot be merged unless tests pass**

If user exposed functionality or configuration variables are added/changed:
  - [x] Documentation added/updated in [home-assistant.github.io](https://github.com/home-assistant/home-assistant.github.io)

If the code communicates with devices, web services, or third-party tools:
  - [x] New dependencies have been added to the `REQUIREMENTS` variable ([example][ex-requir]).
  - [x] New dependencies are only imported inside functions that use them ([example][ex-import]).
  - [x] New or updated dependencies have been added to `requirements_all.txt` by running `script/gen_requirements_all.py`.
  - [x] New files were added to `.coveragerc`.

[ex-requir]: https://github.com/home-assistant/home-assistant/blob/dev/homeassistant/components/keyboard.py#L14
[ex-import]: https://github.com/home-assistant/home-assistant/blob/dev/homeassistant/components/keyboard.py#L54

peternijssen committed:
add spider thermostats

peternijssen commented on homeassistant/components/climate/spider.py:
```diff
@@ -0,0 +1,149 @@
+            _LOGGER.error("No data from the Itho Daalderop API")
+            return
+
+        for prop in self.thermostat['properties']:
```
I wonder if I should set all the properties here or that I should remove the properties and put the logic in the get methods.

MartinHjelmare commented on homeassistant/components/climate/spider.py:
```diff
@@ -0,0 +1,151 @@
+    """Set up the Spider thermostat."""
+    from itho_daalderop_api import IthoDaalderop_API, UnauthorizedException
+
+    username = config.get(CONF_USERNAME)
```
Use `dict[key]` for required keys in the config schema.

peternijssen replied:
Done. Moved to hub/component

MartinHjelmare commented on homeassistant/components/climate/spider.py:
```diff
@@ -0,0 +1,151 @@
+        api = IthoDaalderop_API(username, password)
+
+    except UnauthorizedException:
+        _LOGGER.exception("Can't connect to the Itho Daalderop API")
```
Please use error level.

peternijssen replied:
done

MartinHjelmare commented on homeassistant/components/climate/spider.py:
```diff
@@ -0,0 +1,151 @@
+    @property
+    def supported_features(self):
+        """Return the list of supported features."""
+        supported = (SUPPORT_TARGET_TEMPERATURE)
```
Just return the constant here.

MartinHjelmare replied:
Is it not possible to change the operation mode?

peternijssen replied:
operation mode can only be changed on the master. Added this now.

MartinHjelmare commented on homeassistant/components/climate/spider.py:
```diff
@@ -0,0 +1,151 @@
+
+_LOGGER = logging.getLogger(__name__)
+
+MIN_TIME_BETWEEN_UPDATES = timedelta(minutes=15)
```
Is this throttling time the absolute min time to not be blocked by the api? If you just want to set the scan interval, define the `SCAN_INTERVAL` constant at the module level.

With this throttle time we won't get state feedback during service calls if they happen more often than every 15 min. That seems likely to happen, when changing some settings.

peternijssen replied:
I am not sure if throttling is needed. Trying currently without to see if the API doesn't kick me out.

MartinHjelmare commented on homeassistant/components/climate/spider.py:
```diff
@@ -0,0 +1,151 @@
+        if temperature is None:
+            return
+
+        self._target_temperature = temperature
```
It's better to let `update` update the instance attributes. `update` will be called immediately after each actuator call during service calls.

peternijssen replied:
done

MartinHjelmare reviewed (commented):
Looks good. Some comments and questions?

peternijssen commented:
@MartinHjelmare I plan on also adding the smart powerplugs of the Spider system soon. I think it's wise if I change this PR to a hub + climate and then later on create a PR with the plugs. Or what is your preference? Because I guess accepting this, will lead to a BC when switching to a "hub" to add the plugs.

MartinHjelmare commented:
Yeah, it's better to start with hub + platform directly if you know that more platforms are incoming.

MartinHjelmare commented on homeassistant/components/climate/spider.py:
```diff
@@ -0,0 +1,143 @@
+from homeassistant.components.climate import (
+    ATTR_TEMPERATURE, STATE_COOL, STATE_HEAT, STATE_IDLE,
+    SUPPORT_TARGET_TEMPERATURE, SUPPORT_OPERATION_MODE, ClimateDevice)
+from homeassistant.components.spider import DOMAIN
```
This platform belongs to the climate domain. So we're not allowed to import another domain as `DOMAIN`. Import the spider domain like so:
```py
from homeassistant.components.spider import DOMAIN as SPIDER_DOMAIN
```

MartinHjelmare commented on homeassistant/components/climate/spider.py:
```diff
@@ -0,0 +1,143 @@
+        """Return the list of supported features."""
+        supports = SUPPORT_TARGET_TEMPERATURE
+
+        if self._master is True:
```
```py
if self._master:
```

MartinHjelmare commented on homeassistant/components/climate/spider.py:
```diff
@@ -0,0 +1,143 @@
+        self._min_temp = None
+        self._max_temp = None
+        self._operation = STATE_IDLE
+        self._operation_list = [
```
This can be extracted to a constant at module level.

MartinHjelmare commented on homeassistant/components/climate/spider.py:
```diff
@@ -0,0 +1,143 @@
+
+    def set_operation_mode(self, operation_mode):
+        """Set new target operation mode."""
+        self.client.set_operation_mode(self._thermostat, operation_mode)
```
Do we need to translate between home assistant operation mode and device mode?

peternijssen replied:
heat = Heat
cool = Cool

The API doesn't allow everything lowercased, so I found it better to capitalize the first character within the API lib itself. So no translation needed here.

MartinHjelmare reviewed (commented):
If we don't need any platform specific config options it's better to set up the platforms via discovery from the component. Use the discovery helper function `load_platform` to do this. If needed we can populate the  discovery info dict passed to the platform as an argument to `load_platform`.

https://developers.home-assistant.io/docs/en/creating_component_generic_discovery.html

MartinHjelmare commented:
Please don't squash commits after review has started. Squashing makes it harder to follow changes. We'll squash when merging via github.

peternijssen commented:
Ah, cool. Thank you for telling me.

peternijssen committed:
Added load_platform. Added operation dictionary. Minor improvements

MartinHjelmare commented on homeassistant/components/climate/spider.py:
```diff
@@ -0,0 +1,147 @@
+                self._max_temp = float(prop['max'])
+            if prop['id'] == 'OperationMode':
+                self._master = True
+                self._operation = SPIDER_STATE_TO_HA[prop['status']]
```
Now we don't catch idle state.

peternijssen replied:
Correct. In their web interface there is no such thing as idle. Either it's heating (even when the set temperature is lower then the actual temperature) or cooling (even when your boiler can't cool, the button is present to set it to cool). I added this as a note to the docs.

MartinHjelmare commented on homeassistant/components/spider.py:
```diff
@@ -0,0 +1,54 @@
+            'thermostats': api.get_thermostats()
+        }
+
+        load_platform(hass, 'climate', DOMAIN)
```
Maybe iterate over `SPIDER_COMPONENTS`?

peternijssen committed:
loop over spider components for load_platform

MartinHjelmare commented on homeassistant/components/spider.py:
```diff
@@ -0,0 +1,55 @@
+        }
+
+        for component in SPIDER_COMPONENTS:
+            load_platform(hass, component, DOMAIN)
```
Add an empty dict as discovery_info.

MartinHjelmare commented on homeassistant/components/climate/spider.py:
```diff
@@ -0,0 +1,147 @@
+
+def setup_platform(hass, config, add_devices, discovery_info=None):
+    """Set up the Spider thermostat."""
+    thermostats = hass.data[SPIDER_DOMAIN]['thermostats']
```
Add a check if discovery_info is None and return if so. We only want to set up the platform if it's the component that loaded the platform. Ie not via platform config.

MartinHjelmare commented on homeassistant/components/climate/spider.py:
```diff
@@ -0,0 +1,147 @@
+    api = hass.data[SPIDER_DOMAIN]['controller']
+    for thermostat in thermostats:
+        add_devices(
+            [SpiderThermostat(api, thermostat['id'], thermostat['name'])],
```
Add all devices in one call to `add_devices`.

peternijssen replied:
I had a look at the Tesla component. Not sure if this was what you had in mind.

peternijssen committed:
added empty dict to load_platform. changed add_devices

peternijssen committed:
moved logic to the API

peternijssen commented:
@MartinHjelmare I moved forward meanwhile. I renamed and changed the API library, so HA has less "logic" to do and more is handled within the API itself. I believe this leads to way cleaner code in HA.

Meanwhile I also have the switch component standing by, but I didn't want to overload you yet with files. 

Let me know what is needed to get this merged, so I can create the PR for the switch component

(PS: How can I buy you a beer for your continuous effort in checking my PR and everyone else his PR for this great project?)

peternijssen committed:
fix requirements_all.txt

peternijssen committed:
minor code improvements

MartinHjelmare commented on homeassistant/components/climate/spider.py:
```diff
@@ -0,0 +1,140 @@
+    @property
+    def current_temperature(self):
+        """Return the current temperature."""
+        return self.thermostat.current_temperature
```
Does this attribute (or other) access involve I/O or is it just returning a cached value? I/O is not permitted in entity properties, since they are evaluated in the event loop.

peternijssen replied:
the object "thermostat" is cached within this class itself. It depends on this object in the library;
https://github.com/peternijssen/spiderpy/blob/master/spiderpy/devices/thermostat.py

I guess because of the for loop within that class, I should cache it either in HA or the library?

MartinHjelmare replied:
That looks fine. It's just parsing a json dict. Great!

MartinHjelmare replied:
You could add some caching in the library to avoid unnecessary for loops, but it's not required.

peternijssen replied:
Agree. I will take care of that in the next version.

MartinHjelmare reviewed (commented):
Looks good. One question.

MartinHjelmare commented:
No need for a free beer. :smile: This is my hobby. Thanks for the feedback!
